File: internal/node/middleware.go
package node

import (
	"context"
	"net"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/prometheus/client_golang/prometheus"
)

// RateLimiter implements a simple token bucket rate limiter per IP
type RateLimiter struct {
	buckets map[string]*tokenBucket
	mutex   sync.RWMutex
	rate    int           // requests per second
	burst   int           // max burst size
	cleanup chan struct{}
}

type tokenBucket struct {
	tokens    int
	lastRefill time.Time
	mutex     sync.Mutex
}

func NewRateLimiter(rate, burst int) *RateLimiter {
	rl := &RateLimiter{
		buckets: make(map[string]*tokenBucket),
		rate:    rate,
		burst:   burst,
		cleanup: make(chan struct{}),
	}
	
	go rl.cleanupStaleEntries()
	return rl
}

func (rl *RateLimiter) Allow(ip string) bool {
	rl.mutex.Lock()
	bucket, exists := rl.buckets[ip]
	if !exists {
		bucket = &tokenBucket{
			tokens:     rl.burst,
			lastRefill: time.Now(),
		}
		rl.buckets[ip] = bucket
	}
	rl.mutex.Unlock()
	
	bucket.mutex.Lock()
	defer bucket.mutex.Unlock()
	
	// Refill tokens based on time elapsed
	now := time.Now()
	elapsed := now.Sub(bucket.lastRefill)
	tokensToAdd := int(elapsed.Seconds() * float64(rl.rate))
	
	if tokensToAdd > 0 {
		bucket.tokens += tokensToAdd
		if bucket.tokens > rl.burst {
			bucket.tokens = rl.burst
		}
		bucket.lastRefill = now
	}
	
	if bucket.tokens > 0 {
		bucket.tokens--
		return true
	}
	
	return false
}

func (rl *RateLimiter) cleanupStaleEntries() {
	ticker := time.NewTicker(5 * time.Minute)
	defer ticker.Stop()
	
	for {
		select {
		case <-ticker.C:
			rl.mutex.Lock()
			cutoff := time.Now().Add(-10 * time.Minute)
			for ip, bucket := range rl.buckets {
				bucket.mutex.Lock()
				if bucket.lastRefill.Before(cutoff) {
					delete(rl.buckets, ip)
				}
				bucket.mutex.Unlock()
			}
			rl.mutex.Unlock()
		case <-rl.cleanup:
			return
		}
	}
}

func (rl *RateLimiter) Close() {
	close(rl.cleanup)
}

// SecurityMiddleware provides various security features
type SecurityMiddleware struct {
	rateLimiter    *RateLimiter
	maxRequestSize int64
	metrics        *SecurityMetrics
}

type SecurityMetrics struct {
	rateLimitedRequests   prometheus.Counter
	oversizedRequests     prometheus.Counter
	suspiciousRequests    prometheus.Counter
}

func NewSecurityMiddleware(rateLimit, burst int, maxRequestSize int64) *SecurityMiddleware {
	metrics := &SecurityMetrics{
		rateLimitedRequests: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "repram_rate_limited_requests_total",
			Help: "Total number of rate-limited requests",
		}),
		oversizedRequests: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "repram_oversized_requests_total",
			Help: "Total number of oversized requests rejected",
		}),
		suspiciousRequests: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "repram_suspicious_requests_total",
			Help: "Total number of suspicious requests detected",
		}),
	}
	
	prometheus.MustRegister(
		metrics.rateLimitedRequests,
		metrics.oversizedRequests,
		metrics.suspiciousRequests,
	)
	
	return &SecurityMiddleware{
		rateLimiter:    NewRateLimiter(rateLimit, burst),
		maxRequestSize: maxRequestSize,
		metrics:        metrics,
	}
}

func (sm *SecurityMiddleware) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Apply security headers
		sm.applySecurityHeaders(w)
		
		// Check rate limiting
		clientIP := sm.getClientIP(r)
		if !sm.rateLimiter.Allow(clientIP) {
			sm.metrics.rateLimitedRequests.Inc()
			http.Error(w, "Rate limit exceeded", http.StatusTooManyRequests)
			return
		}
		
		// Check request size
		if r.ContentLength > sm.maxRequestSize {
			sm.metrics.oversizedRequests.Inc()
			http.Error(w, "Request too large", http.StatusRequestEntityTooLarge)
			return
		}
		
		// Check for suspicious patterns
		if sm.isSuspiciousRequest(r) {
			sm.metrics.suspiciousRequests.Inc()
			http.Error(w, "Forbidden", http.StatusForbidden)
			return
		}
		
		// Add security context
		ctx := r.Context()
		ctx = context.WithValue(ctx, "client_ip", clientIP)
		r = r.WithContext(ctx)
		
		next.ServeHTTP(w, r)
	})
}

func (sm *SecurityMiddleware) applySecurityHeaders(w http.ResponseWriter) {
	// Basic security headers
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.Header().Set("X-Frame-Options", "DENY")
	w.Header().Set("X-XSS-Protection", "1; mode=block")
	w.Header().Set("Referrer-Policy", "strict-origin-when-cross-origin")
	w.Header().Set("Content-Security-Policy", "default-src 'self'")
	
	// REPRAM-specific headers
	w.Header().Set("X-REPRAM-Node", "1.0.0")
}

func (sm *SecurityMiddleware) getClientIP(r *http.Request) string {
	// Check X-Forwarded-For header first (for proxies)
	xff := r.Header.Get("X-Forwarded-For")
	if xff != "" {
		// Take the first IP in the chain
		ips := strings.Split(xff, ",")
		if len(ips) > 0 {
			return strings.TrimSpace(ips[0])
		}
	}
	
	// Check X-Real-IP header
	xri := r.Header.Get("X-Real-IP")
	if xri != "" {
		return xri
	}
	
	// Fall back to RemoteAddr
	ip, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		return r.RemoteAddr
	}
	return ip
}

func (sm *SecurityMiddleware) isSuspiciousRequest(r *http.Request) bool {
	userAgent := r.Header.Get("User-Agent")
	
	// Block common bot/scanner user agents
	suspiciousUAs := []string{
		"sqlmap",
		"nikto",
		"nmap",
		"masscan",
		"gobuster",
		"dirbuster",
		"<script",
		"python-requests", // Block basic scripts (unless legitimate usage)
	}
	
	userAgentLower := strings.ToLower(userAgent)
	for _, suspicious := range suspiciousUAs {
		if strings.Contains(userAgentLower, suspicious) {
			return true
		}
	}
	
	// Check for SQL injection patterns in URL
	url := strings.ToLower(r.URL.String())
	sqlPatterns := []string{
		"union", "select", "insert", "delete", "drop", "exec",
		"script", "alert", "onerror", "onload",
		"../", "..\\", "/etc/passwd", "/proc/",
	}
	
	for _, pattern := range sqlPatterns {
		if strings.Contains(url, pattern) {
			return true
		}
	}
	
	return false
}

func (sm *SecurityMiddleware) Close() {
	if sm.rateLimiter != nil {
		sm.rateLimiter.Close()
	}
}

// Request size limiting middleware
func MaxRequestSizeMiddleware(maxSize int64) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if r.ContentLength > maxSize {
				http.Error(w, "Request too large", http.StatusRequestEntityTooLarge)
				return
			}
			
			// Also set a limit on the request body reader
			r.Body = http.MaxBytesReader(w, r.Body, maxSize)
			next.ServeHTTP(w, r)
		})
	}
}

// Timeout middleware to prevent slow loris attacks
func TimeoutMiddleware(timeout time.Duration) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.TimeoutHandler(next, timeout, "Request timeout")
	}
}

File: repram-mcp/src/node/middleware.ts
/**
 * Security middleware — rate limiting, scanner detection, request guards.
 *
 * Port of internal/node/middleware.go. No mutexes needed in single-threaded
 * Node.js. These are standalone utilities consumed by the HTTP server (P2b).
 */

import type { IncomingMessage, ServerResponse } from "node:http";

// ─── Rate Limiter ────────────────────────────────────────────────────

interface TokenBucket {
  tokens: number;
  lastRefill: number; // Date.now()
}

export class RateLimiter {
  private buckets = new Map<string, TokenBucket>();
  private rate: number;       // tokens per second
  private burst: number;      // max tokens
  private cleanupTimer: ReturnType<typeof setInterval> | null;

  constructor(ratePerSecond: number, burst: number) {
    this.rate = ratePerSecond;
    this.burst = burst;
    // Evict stale buckets every 5 minutes (matching Go)
    this.cleanupTimer = setInterval(() => this.cleanupStale(), 5 * 60_000);
  }

  allow(ip: string): boolean {
    let bucket = this.buckets.get(ip);
    if (!bucket) {
      bucket = { tokens: this.burst, lastRefill: Date.now() };
      this.buckets.set(ip, bucket);
    }

    // Refill based on elapsed time
    const now = Date.now();
    const elapsedSeconds = (now - bucket.lastRefill) / 1000;
    const tokensToAdd = Math.floor(elapsedSeconds * this.rate);

    if (tokensToAdd > 0) {
      bucket.tokens = Math.min(this.burst, bucket.tokens + tokensToAdd);
      bucket.lastRefill = now;
    }

    if (bucket.tokens > 0) {
      bucket.tokens--;
      return true;
    }

    return false;
  }

  /** Number of tracked IPs (exposed for testing). */
  get bucketCount(): number {
    return this.buckets.size;
  }

  close(): void {
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = null;
    }
  }

  private cleanupStale(): void {
    const cutoff = Date.now() - 10 * 60_000; // 10 minutes
    for (const [ip, bucket] of this.buckets) {
      if (bucket.lastRefill < cutoff) {
        this.buckets.delete(ip);
      }
    }
  }
}

// ─── Scanner Detection ───────────────────────────────────────────────

/**
 * Known vulnerability scanner user-agents. Only scanner-specific tools —
 * not general-purpose HTTP libraries. REPRAM is permissionless by design;
 * python-requests, curl, etc. are legitimate clients.
 */
const SCANNER_USER_AGENTS = [
  "sqlmap",
  "nikto",
  "nmap",
  "masscan",
  "gobuster",
  "dirbuster",
];

export function isScannerRequest(userAgent: string): boolean {
  const lower = userAgent.toLowerCase();
  return SCANNER_USER_AGENTS.some((scanner) => lower.includes(scanner));
}

// ─── Client IP Extraction ────────────────────────────────────────────

/**
 * Extract the client IP from an incoming HTTP request.
 * Only trusts proxy headers (X-Forwarded-For, X-Real-IP) when trustProxy
 * is true. Otherwise uses the direct socket address.
 */
export function getClientIP(
  req: IncomingMessage,
  trustProxy: boolean,
): string {
  if (trustProxy) {
    const xff = req.headers["x-forwarded-for"];
    if (xff) {
      const first = (Array.isArray(xff) ? xff[0] : xff).split(",")[0].trim();
      if (first) return first;
    }
    const xri = req.headers["x-real-ip"];
    if (xri) {
      return Array.isArray(xri) ? xri[0] : xri;
    }
  }

  // Direct connection IP
  return req.socket?.remoteAddress ?? "unknown";
}

// ─── Security Headers ────────────────────────────────────────────────

export function applySecurityHeaders(res: ServerResponse): void {
  res.setHeader("X-Content-Type-Options", "nosniff");
  res.setHeader("X-Frame-Options", "DENY");
  res.setHeader("X-XSS-Protection", "1; mode=block");
  res.setHeader("Referrer-Policy", "strict-origin-when-cross-origin");
  res.setHeader("Content-Security-Policy", "default-src 'self'");
  res.setHeader("X-REPRAM-Node", "2.0.0");
}

// ─── CORS ────────────────────────────────────────────────────────────

/**
 * Apply CORS headers. REPRAM accepts any origin by design — it's a
 * permissionless network primitive.
 */
export function applyCorsHeaders(
  req: IncomingMessage,
  res: ServerResponse,
): void {
  const origin = req.headers.origin;
  if (origin) {
    res.setHeader("Access-Control-Allow-Origin", origin);
  }
  res.setHeader("Access-Control-Allow-Methods", "GET, PUT, POST, HEAD, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type, X-TTL, X-Repram-Signature");
  res.setHeader("Access-Control-Max-Age", "3600");
}

// ─── Composed SecurityMiddleware ─────────────────────────────────────

export interface SecurityMiddlewareOptions {
  rateLimit: number;
  burst: number;
  maxRequestSize: number;
  trustProxy: boolean;
}

export class SecurityMiddleware {
  readonly rateLimiter: RateLimiter;
  readonly maxRequestSize: number;
  private trustProxy: boolean;

  constructor(options: SecurityMiddlewareOptions) {
    this.rateLimiter = new RateLimiter(options.rateLimit, options.burst);
    this.maxRequestSize = options.maxRequestSize;
    this.trustProxy = options.trustProxy;
  }

  /**
   * Run all security checks. Returns the client IP if allowed,
   * or null if the request was rejected (response already sent).
   */
  check(req: IncomingMessage, res: ServerResponse): string | null {
    applySecurityHeaders(res);

    const clientIP = getClientIP(req, this.trustProxy);

    if (!this.rateLimiter.allow(clientIP)) {
      res.writeHead(429, { "Content-Type": "text/plain" });
      res.end("Rate limit exceeded");
      return null;
    }

    const contentLength = parseInt(req.headers["content-length"] ?? "0", 10);
    if (contentLength > this.maxRequestSize) {
      res.writeHead(413, { "Content-Type": "text/plain" });
      res.end("Request too large");
      return null;
    }

    const userAgent = req.headers["user-agent"] ?? "";
    if (isScannerRequest(userAgent)) {
      res.writeHead(403, { "Content-Type": "text/plain" });
      res.end("Forbidden");
      return null;
    }

    return clientIP;
  }

  close(): void {
    this.rateLimiter.close();
  }
}
